### PR TITLE
Add option --configfile

### PR DESCRIPTION
**Pull request by fast90**
### Before submitting a _pull request_ make sure you have:
- [ x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](htt2ps://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x ] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
### What is the purpose of your _pull request_?
- [x ] New feature

---

A configfile can now be passed to youtube-dl. Probably closes (#6745)?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -243,12 +247,11 @@ which means you can modify it, redistribute it or use it however you like.
...
-                                     $XDG_CACHE_HOME/youtube-dl or
-                                     ~/.cache/youtube-dl . At the moment, only
-                                     YouTube player files (for videos with
-                                     obfuscated signatures) are cached, but that
-                                     may change.
+                                     permanently. By default $XDG_CACHE_HOME
+                                     /youtube-dl or ~/.cache/youtube-dl . At the
+                                     moment, only YouTube player files (for
+                                     videos with obfuscated signatures) are
+                                     cached, but that may change.
```
Comment (yan12125) [inline README.md]: Please leave README.md untouched. Different pandoc versions lead to different results, so it's preferred to update it centrally.

File: youtube_dl/options.py [hunk 2]
```diff
@@ -848,19 +854,31 @@ def compat_conf(conf):
...
+        opts, args = parser.parse_args(command_line_conf)
 
         if '--ignore-config' in command_line_conf:
             system_conf = []
             user_conf = []
+        elif '--configfile' in command_line_conf:
+            if not os.path.isfile(opts.configfile):
+                write_string('configfile not found, using default configfiles instead \n')
+                system_conf = _readOptions('/etc/youtube-dl.conf')
+                user_conf = _readUserConf()
```
Comment (yan12125) [inline youtube_dl/options.py]: It's better to fail instead of fallback.

File: youtube_dl/options.py [hunk 3]
```diff
@@ -848,19 +854,31 @@ def compat_conf(conf):
...
+            else:
+                user_conf = []
+                system_conf = _readOptions(opts.configfile)
```
Comment (yan12125) [inline youtube_dl/options.py]: Config files specified via the command line sounds like a user configuration for me.

File: youtube_dl/options.py [hunk 4]
```diff
@@ -179,6 +179,12 @@ def _scrub_eq(o):
         'Do not read the user configuration in ~/.config/youtube-dl/config '
         '(%APPDATA%/youtube-dl/config.txt on Windows)')
     general.add_option(
+        '--configfile',
+        dest='configfile', metavar='CONFIGFILE', type=str,
+        help='Do not read default configuration files found in '
+        '/etc/youtube-dl.conf and ~/.config/youtube-dl/config. '
```
Comment (dstftw) [inline youtube_dl/options.py]: Is's absolutely pointless to clarify paths here. youtube-dl may be running on Window host where these paths make no sense at all.

File: youtube_dl/options.py [hunk 5]
```diff
@@ -179,6 +179,12 @@ def _scrub_eq(o):
         'Do not read the user configuration in ~/.config/youtube-dl/config '
         '(%APPDATA%/youtube-dl/config.txt on Windows)')
     general.add_option(
+        '--configfile',
+        dest='configfile', metavar='CONFIGFILE', type=str,
```
Comment (dstftw) [inline youtube_dl/options.py]: metavar should be `FILE`. `type` is already string be default.

**Comment by fast90** [comment]
Reverted changes to README.md, raise an error if configfile not found, fixed helptext.

File: youtube_dl/options.py [hunk 6]
```diff
@@ -179,6 +179,11 @@ def _scrub_eq(o):
         'Do not read the user configuration in ~/.config/youtube-dl/config '
         '(%APPDATA%/youtube-dl/config.txt on Windows)')
     general.add_option(
+        '--configfile',
+        dest='configfile', metavar='FILE',
+        help='Do not read default configuration files. '
+        'Use The file found in FILE instead')
```
Comment (dstftw) [inline youtube_dl/options.py]: Just `File to read configuration from`.

File: youtube_dl/options.py [hunk 7]
```diff
@@ -848,19 +853,29 @@ def compat_conf(conf):
...
 
         command_line_conf = compat_conf(sys.argv[1:])
+        opts, args = parser.parse_args(command_line_conf)
 
         if '--ignore-config' in command_line_conf:
             system_conf = []
             user_conf = []
+        elif '--configfile' in command_line_conf:
+            if not os.path.isfile(opts.configfile):
+                raise parser.error('Config file {} not found.'.format(opts.configfile))
```
Comment (dstftw) [inline youtube_dl/options.py]: `{}` won't work in python 2.6.

File: youtube_dl/options.py [hunk 8]
```diff
@@ -179,6 +179,11 @@ def _scrub_eq(o):
         'Do not read the user configuration in ~/.config/youtube-dl/config '
         '(%APPDATA%/youtube-dl/config.txt on Windows)')
     general.add_option(
+        '--configfile',
+
```
Comment (yan12125) [inline youtube_dl/options.py]: Something wrong?

File: youtube_dl/options.py [hunk 9]
```diff
@@ -179,6 +179,11 @@ def _scrub_eq(o):
         'Do not read the user configuration in ~/.config/youtube-dl/config '
         '(%APPDATA%/youtube-dl/config.txt on Windows)')
     general.add_option(
+        '--configfile',
```
Comment (dstftw) [inline youtube_dl/options.py]: Should be `--config-file` or `--config`.

File: youtube_dl/options.py [hunk 10]
```diff
@@ -848,19 +852,29 @@ def compat_conf(conf):
...
 
         command_line_conf = compat_conf(sys.argv[1:])
+        opts, args = parser.parse_args(command_line_conf)
 
         if '--ignore-config' in command_line_conf:
             system_conf = []
             user_conf = []
+        elif '--config-file' in command_line_conf:
+            if not os.path.isfile(opts.configfile):
+                raise parser.error('Config file {0} not found.'.format(opts.configfile))
```
Comment (yan12125) [inline youtube_dl/options.py]: ```parser.error``` is a function, not an exception class.

Comment (fast90) [inline youtube_dl/options.py]: What exception class would you use?

Comment (yan12125) [inline youtube_dl/options.py]: ```parser.error``` calls ```sys.exit```, so simply ```parser.error('....')``` (without ```raise```) should work.

**Comment by fast90** [comment]
Anything missing?

**Comment by yan12125** [comment]
Please squash all commits into a single one.

**Comment by fast90** [comment]
Done

**Comment by fast90** [comment]
@dstftw Could you please approve? Thank you.

**Comment by fast90** [comment]
Anyone?

**Pull request closed without merging**